### DISCUSSION TITLE
Seeking advice on converting nested Tibble exported from R to suitable dataframe for sktime

### DISCUSSION THREAD

**Discussion by jmwhyte** [alan-turing-institute/sktime#1890] [Q&A]
R users looking to apply sktime might find some wrangling is necessary.
There are two issues (which I will illustrate with an example shortly):
1. Exporting a dataframe from R and importing it to Python. (I tried to do everything in a Jupyter notebook, and R's reticulate crashed the kernel).
2. Handling the conversion to a suitable form. For example, in R, using tidyr::nest makes a tibble (type of data frame) that can have some elements which are also a tibble. For each such entry, there is a row number, and the numerical values of interest under some heading. This heading is problematic, as it seems to upset sktime, which is probably a consequence of the data not being in the correct form.

Example follows:

### Step 1: Preparing the data object. 
(**NOTE:** I am also investigating an approach using rpy2 so that the whole process can be undertaken in Jupyter, which will simplify this part. There's currently a problem with `tidyr::nest()` that I'm investigating.)
 
In Rstudio

```{r}
# This is an "r" cell
library(dplyr)
library(tibble)
# For simplicity, let's only deal with a univariate series for each species of iris
# Create a nested tibble:
nested.iris <- select(iris,c(Sepal.Length,Species)) %>% tidyr::nest(data = !Species)
# Consider the R object and one entry of it in column two 
print(nested.iris)
 print(nested.iris[1,2][[1]])
 attributes(nested.iris[1,2])
# Setup the objects for sktime
rX <- nested.iris
print(head(rX))
```

```{python}
# This is a "python" cell
import numpy
import pandas
import os

# For saving python objects
import pickle

# Keep the existing column names from the R tibble
X = pandas.DataFrame(r.rX)

with open('Iris_data_example.pydata','wb') as out_file:
  pickle.dump(X, out_file)
  
exit()
```

### Step 2: object processing

In a Jupyter (ipython) notebook :
Load the object created by R

import pickle
with open('Iris_data_example.pydata', 'rb') as in_file:
    X = pickle.load(in_file)
% X is <class 'pandas.core.frame.DataFrame'>, shape (3, 2), columns are: Index(['Species', 'data'], dtype='object')
% An element of the data column, e.g. X['data'][0] is <class 'pandas.core.frame.DataFrame'>, shape (50, 1), columns are: Index(['Sepal.Length'], dtype='object')
% Further, within a dataframe element, consider the column of interest:
% print(type(X['data'][0]['Sepal.Length']))
% <class 'pandas.core.series.Series'>

% Consider some features of an element of X which houses a series
print(X['data'][1])
print(X['data'][1]['Sepal.Length'])

% Put each of the elements which is a series into a suitable form for sktime
import pandas as pd
import numpy as np

Xfinal = pd.DataFrame(index=np.arange(X.shape[0]), columns = X.columns)

for i in range(0,X.shape[0]):
    for j in X.columns[1:]:
        Xfinal[j][i] = pd.Series(X[j][i]['Sepal.Length'])

Drop the non-time series column 
Xfinal = Xfinal.drop('Species',axis=1)

This is good enough to go into sktime and obtain a (nonsense) result.
However, this is all quite amateurish, and I'm sure someone somewhere (other than myself) will benefit from guidance on how to run this process more efficiently.

**Comment by lmmentel** [comment]
Thanks for starting the thread on this. It would be easier to help out if you would be able to post a minimal example that we could reason about. Would you be able to put one together?

**Reply by jmwhyte** [reply]
I will certainly put some time into a MWE tomorrow and report back.

**Reply by jmwhyte** [reply]
@lmmentel I've put up a clunky effort that should get the discussion going. It was a little different to my original example as it seems that the column label could be removed with .values, but I hope this is adequate for a first effort.

**Reply by fkiraly** [reply]
hm, could you perhaps be explicit what the type/format of `X` is once you load this from pickle? Apologies, I don't have R running on my local setup I use for `sktime` dev.

**Reply by jmwhyte** [reply]
Thanks for the comment, I've added some lines below the pickle load command that address your request.

**Reply by lmmentel** [reply]
Thanks for the updates, it makes it clearer now. Am I correct that you are passing a variable form R to python between cells? Independently of that I would encourage you to serialize the R data to a language agnostic format such as csv/tsv, parquet, json, etc and the load that data in python. In case that wouldn't work for you could you explain why?

**Reply by jmwhyte** [reply]
@lmmentel  wrote: "Am I correct that you are passing a variable form R to python between cells?" Well yes ... but in an indirect way because there's a pickle.dump followed by a pickle.load, which seems very inefficient, but I couldn't find another way to get the object from R into python quickly. (It didn't occur to me that something like .csv would be a better format for the data!)

However, rather than save/load I'm trying to get this to **all** happen in Python to streamline this example (at least). This "Stack Overflow" question shows that I made some progress (using `rpy2`), but making the nested dataframe in R using the `rtidy::nest()` is the sticking point that I still need to resolve: [Stack Overflow rpy2/nest question](https://stackoverflow.com/questions/70766541/how-to-use-rpy2-v3-4-5-with-rs-tidyr-nest)

Also: "Independently of that I would encourage you to serialize the R data to a language agnostic format such as csv/tsv, parquet, json, etc and the load that data in python. In case that wouldn't work for you could you explain why?"

From in %%R cell in Jupyter, I tried two different ways of exporting X (see its appearance in R output below) to .csv

> # A tibble: 3 x 2
>   Species    data             
>   <fct>      <list>           
> 1 setosa     <tibble [50 × 1]>
> 2 versicolor <tibble [50 × 1]>
> 3 virginica  <tibble [50 × 1]>
> 

> write.table(rX,"rX.csv", sep=",") %fails
R[write to console]: Error in write.table(rX, "rX.csv", sep = ",") : 
  unimplemented type 'list' in 'EncodeElement'

> write.csv(rX,"rX.csv")   %this also fails
R[write to console]: Error in write.table(rX, "rX.csv", col.names = NA, sep = ",", dec = ".",  : 
  unimplemented type 'list' in 'EncodeElement'

So at this stage, the nested dataframe seems to cause problems for .csv export.

 I would need to check on other export formats suggested, as I'm not familiar with those.

**Comment by fkiraly** [comment]
Ah, thanks for the example.

The object `X` in question is *almost* in one of the `sktime` supported formats - the `nested_univ` format. The difference is that the latter contains `pandas.Series` in the cells, not `pandas.DataFrame`. 

Since the `pd.DataFrame` inside the cells have only one column, you could do
`Xfinal = X.applymap(lambda x: x.iloc[:, 0])`
(to convert them to `pd.Series` manually)
and that should be in the `nested_univ` format.

From there, you can use `datatypes.convert_to` to convert to any other `sktime` compatible format

If you want to be systematic, you could add a new format to `datatypes` (but that may be overkill if this input format of nested `pd.DataFrame` is rarely encountered).

**Reply by jmwhyte** [reply]
@fkiraly I tried your suggestion for Χfinal and then saw an error:
"AttributeError: 'str' object has no attribute 'iloc'"
Perhaps there's some other step required (e.g. package I needed to load)? Can you please advise?

**Reply by fkiraly** [reply]
hm, maybe then I didn't get the input format correctly. I based this purely on the printout.

One plausible reason for this is that you are applying the above to `X_final`, not `X`. My suggestion is to do this with `X`, right after the line `X = pickle.load(in_file)`, not with `X_final` that you get at the end.

If that doesn't work: would you perhaps be able to post a single cell of python code that creates `X` as in the line `X = pickle.load(in_file)`, without reliance on ap pickle (i.e., create an object of similar format manually). That way, I can play around much more easily and find sth (or code sth) for you.

**Reply by jmwhyte** [reply]
@fkiraly , I definitely applied your suggestion to X, not Xfinal. I've added some more detail on what the series within each cell of the dataframe looks like: see "print(type(X['data'][0]['Sepal.Length']))" in the original post. Could you please also clarify your last paragraph above? "ap pickle"?

**Reply by fkiraly** [reply]
sorry, should have said "without reliance on a pickle". I.e., just a few lines of python code that create a container of the same format that `X` is at the start, so I can play around with it.

**Reply by jmwhyte** [reply]
@fkiraly - here's a Python-only approach to getting the nested tibble X (out of R, using rpy2 commands) for subsequent use in Python. This was done with rookie Python knowledge, so some lines might be unnecessary, but it works. I hope this is useful.

```
from rpy2 import robjects
from rpy2.robjects import Formula, Environment
from rpy2.robjects.vectors import IntVector, FloatVector
from rpy2.robjects.lib import grid
from rpy2.robjects.packages import importr, data
from rpy2.rinterface_lib.embedded import RRuntimeError
import warnings

base = importr('base')
# I will take the Iris data set from R, as per the original query
datasets = importr('datasets')

from functools import partial
import rpy2.robjects.lib.tidyr as tidyr
from collections import OrderedDict
from rpy2.robjects.vectors import (StrVector,
                                   IntVector)
 
tidyr = importr('tidyr')
dplyr = importr('dplyr')

iris = data(datasets).fetch('iris')['iris']

from rpy2.robjects.lib.tidyr import DataFrame
from rpy2.robjects import rl

dataf = (
  DataFrame(iris) 
  .select(base.c(1,5))
)

print(dataf.head())

X = tidyr.nest(dataf, data=rl('Sepal.Length'))
print(X.head())
```

**Reply by fkiraly** [reply]
thanks! unfortunately, `rpy2` does not work for me - whenever I try to import it, the kernel gets stuck and returns only "session cannot generate requesets". I tried around but have to cut my time debugging that.

Would you be able to, or is someone else able to, produce code that creates the data frame `X` (or a similar one) without importing `rpy2`?

**Reply by jmwhyte** [reply]
@fkiraly I'll need to think about how to make X. However, maybe this is a quick fix: I've uploaded the file I defined in my original post, "Iris_data_example.pydata" (zipped to meet the upload file format requirements) which contains X. So, you should be able to use this to get to X with the lines from above:
```
import pickle
with open('Iris_data_example.pydata', 'rb') as in_file:
X = pickle.load(in_file)
```
[Iris_data_example.pydata.zip](https://github.com/alan-turing-institute/sktime/files/8008955/Iris_data_example.pydata.zip)

**Reply by fkiraly** [reply]
ok, thanks.

I've understood what was wrong above: you have the label column in the same data frame, you need to remove it. Otherwise all what I said applies.

Btw, you can check compliance with the `sktime` data formats by using `check_is_mtype`. If you want it in the `nested_univ` format, you would write sth like
```python
from sktime.datatypes import check_is_mtype
check_is_mtype(X, "nested_univ", return_metadata=True)
```
This gives the informative error message `obj entries must be pd.Series`, so that's what we must achieve, here by:
* removing the label column (that needs to go in `y`), and
* converting the `pd.DataFrame` to `pd.Series`, let's do this:

```python
X_series = X.loc[:, ["data"]]  # remove the label column
X_series = X_series.applymap(lambda x: x.iloc[:, 0])  # convert `pd.DataFrame` entries to `pd.Series`
```

Now, if you check it again:
```python
check_is_mtype(X_series, "nested_univ", return_metadata=True)
```
all is fine:
```
(True,
 None,
 {'is_univariate': True,
  'n_instances': 3,
  'is_one_series': False,
  'has_nans': False,
  'is_equal_length': True,
  'is_empty': False,
  'is_equally_spaced': True,
  'mtype': 'nested_univ',
  'scitype': 'Panel'})
```
